### PR TITLE
AR association callbacks accept an array

### PR DESCRIPTION
**Pull request by ghiculescu**
See docs @ https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html

image

### CODE REVIEW

File: lib/activerecord/all/activerecord.rbi [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
 
 VariadicUntypedFunction = T.type_alias { Proc }
+AssociationCallback = T.type_alias do
+  # docs in https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html - jump to "Association callbacks"
+  T.nilable(
+    T.any(
+      Symbol, # reference to a method
+      String, # reference to a method? not clear: no string examples in docs
+      T.proc.void, # a proc that contains the callback
+      T::Array[T.any(Symbol, Proc)] # multiple callbacks
```
Comment (ghiculescu) [inline lib/activerecord/all/activerecord.rbi]: i dont really understand why `Proc` works here but `T.proc.void` doesn't, [same issue here](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Asig%20%7Breturns(T%3A%3AArray%5BT.any(Symbol%2C%20T.proc.void)%5D)%7D%0Adef%20bar%0A%20%20%5B%3Afoo%2C%20Proc.new%20%7B%20%7D%5D%0Aend%0A%0Asig%20%7Breturns(T%3A%3AArray%5BT.any(Symbol%2C%20Proc)%5D)%7D%0Adef%20bar2%0A%20%20%5B%3Afoo%2C%20Proc.new%20%7B%20%7D%5D%0Aend%0A)

Comment (elliottt) [reply lib/activerecord/all/activerecord.rbi]: Do you get a type error in the rbi file, or does it end up being impossible to use?

Comment (ghiculescu) [reply lib/activerecord/all/activerecord.rbi]:
```
testing activerecord all
activerecord_test.rb:22: Expected T.nilable(T.any(Symbol, String, T.proc.void, T::Array[T.any(T.proc.void, Symbol)])) but found [Symbol(:"foo"), Proc] for argument after_add https://srb.help/7002
    22 |  has_and_belongs_to_many :projects, after_add: [:foo, Proc.new { }]
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    activerecord.rbi:158: Method ActiveRecord::Associations::ClassMethods#has_and_belongs_to_many has specified after_add as T.nilable(T.any(Symbol, String, T.proc.void, T::Array[T.any(T.proc.void, Symbol)]))
     158 |      after_add: AssociationCallback,
                ^^^^^^^^^
  Got [Symbol(:"foo"), Proc] originating from:
    activerecord_test.rb:22:
    22 |  has_and_belongs_to_many :projects, after_add: [:foo, Proc.new { }]
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (elliottt) [reply lib/activerecord/all/activerecord.rbi]: I think the problem is with the use of `Proc.new {}`. If you change that to `lambda {}` or `-> {}` it works fine.

Comment (ghiculescu) [reply lib/activerecord/all/activerecord.rbi]: Ahh gotcha. I just copied the example from the Rails docs. Maybe I should just do `T::Array[T.any(Symbol, Proc, T.proc.void)]`?

Comment (ghiculescu) [reply lib/activerecord/all/activerecord.rbi]: Should we add some more detail to https://sorbet.org/docs/procs about the difference between `Proc` and `T.proc`?

Comment (elliottt) [reply lib/activerecord/all/activerecord.rbi]: Sure! Do you feel like taking an initial pass at that?

Comment (ghiculescu) [reply lib/activerecord/all/activerecord.rbi]: It's on the todo list :) In the meantime I've fixed this one up.

Comment (elliottt) [reply lib/activerecord/all/activerecord.rbi]: Awesome, thank you!

**Pull request merged**